translator/trace/otlp/otlp_to_oc.go

```diff
@@ -0,0 +1,441 @@
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 eventsToOCEvents(events),
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventsToOCEvents(events []*otlptrace.Span_Event) []*octrace.Span_TimeEvent {
+	if len(events) == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvent := &octrace.Span_TimeEvent{
+			Time: unixnanoToTimestamp(event.TimeUnixnano),
+		}
+
+		if isMessageEvent(event.Attributes) {
+			// Handle as a MessageEvent
+			ocEvent.Value = eventAttributesToMessageEvent(event.Attributes)
+		} else {
+			// Handle as an Annotation
+			spanAttributes := attributesToOCSpanAttributes(event.Attributes, event.DroppedAttributesCount)
+			ocEvent.Value = &octrace.Span_TimeEvent_Annotation_{
+				Annotation: &octrace.Span_TimeEvent_Annotation{
+					Description: &octrace.TruncatableString{
+						Value: event.Name,
+					},
+					Attributes: spanAttributes,
+				},
+			}
+		}
+
+		ocEvents = append(ocEvents, ocEvent)
+	}
+	return ocEvents
+}
+
+func isMessageEvent(attributes []*otlpcommon.AttributeKeyValue) bool {
+	for _, attr := range attributes {
+		if attr.Key == ocTimeEventMessageEventType {
+			return true
+		}
+	}
+	return false
+}
+
+func eventAttributesToMessageEvent(attributes []*otlpcommon.AttributeKeyValue) *octrace.Span_TimeEvent_MessageEvent_ {
+	msgEvent := &octrace.Span_TimeEvent_MessageEvent{}
+
+	for _, attr := range attributes {
+		switch attr.Key {
+		case ocTimeEventMessageEventType:
+			msgEvent.Type = octrace.Span_TimeEvent_MessageEvent_Type(octrace.Span_TimeEvent_MessageEvent_Type_value[attr.StringValue])
+		case ocTimeEventMessageEventID:
+			msgEvent.Id = uint64(attr.IntValue)
+		case ocTimeEventMessageEventUSize:
+			msgEvent.UncompressedSize = uint64(attr.IntValue)
+		case ocTimeEventMessageEventCSize:
+			msgEvent.CompressedSize = uint64(attr.IntValue)
+		}
+	}
+
+	return &octrace.Span_TimeEvent_MessageEvent_{
+		MessageEvent: msgEvent,
+	}
+}
+
+func statusToOC(status *otlptrace.Status) *octrace.Status {
+	if status == nil {
+		return nil
+	}
+	return &octrace.Status{
+		Code:    int32(status.Code),
+		Message: status.Message,
+	}
+}
+
+func resourceToOC(resource *otlpresource.Resource) (*occommon.Node, *ocresource.Resource) {
+	if resource == nil {
+		return nil, nil
+	}
+
+	if len(resource.Attributes) == 0 {
+		return nil, nil
+	}
+
+	ocNode := occommon.Node{}
+	ocResource := ocresource.Resource{}
+
+	labels := make(map[string]string, len(resource.Attributes))
+	for _, attr := range resource.Attributes {
+		key := attr.Key
+		val := attributeValueToString(attr)
+
+		switch key {
+		case ocAttributeResourceType:
+			ocResource.Type = val
+		case conventions.AttributeServiceName:
+			ocNode.ServiceInfo.Name = val
+		case ocAttributeProcessStartTime:
+			t, err := time.Parse(time.RFC3339Nano, val)
+			if err != nil {
+				continue
```

Comment:
It is best to set some constant value which indicates that the time is invalid instead of silently dropping the start time.